resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/K8sSubmitOps.scala

```diff
@@ -0,0 +1,178 @@
+
+private[spark] sealed trait K8sSubmitOp extends CommandLineLoggingUtils {
+  def executeOnPod(pName: String, ns: String, gracePeriodSeconds: Option[Long])
+      (implicit client: KubernetesClient): Unit
+  def executeOnGlob(pods: List[Pod], ns: String, gracePeriodSeconds: Option[Long])
+      (implicit client: KubernetesClient): Unit
+}
+
+private[spark] class KillApplication extends K8sSubmitOp  {
+  override def executeOnPod(pName: String, ns: String, gracePeriodSeconds: Option[Long])
+      (implicit client: KubernetesClient): Unit = {
+    val podToDelete = client
+      .pods
+      .inNamespace(ns)
+      .withName(pName)
+      gracePeriodSeconds match {
+        case Some(period) => podToDelete.withGracePeriod(period).delete()
+        case _ => podToDelete.delete()
+      }
+  }
+
+  override def executeOnGlob(pods: List[Pod], ns: String, gracePeriodSeconds: Option[Long])
+      (implicit client: KubernetesClient): Unit = {
+    if (pods.nonEmpty) {
+      pods.foreach { pod => printMessage(s"Deleting driver pod: ${pod.getMetadata.getName}.") }
+      val nClient = client.pods.inNamespace(ns)
+      gracePeriodSeconds match {
+        case Some(period) =>
+          // this is not using the batch api because no option is provided
+          // when using the grace period.
+          pods.foreach { pod =>
+             nClient
+               .withName(pod.getMetadata.getName)
+               .withGracePeriod(period)
+               .delete()
+          }
+        case _ => nClient.delete(pods.asJava)
+      }
+    } else {
+      printMessage("No applications found.")
+    }
+  }
+}
+
+private[spark] class ListStatus extends K8sSubmitOp {
+  override def executeOnPod(pName: String, ns: String, gracePerio: Option[Long])
+      (implicit client: KubernetesClient): Unit = {
+    val pod = client
+      .pods
+      .inNamespace(ns)
+      .withName(pName)
+      .get()
+
+    printMessage("Application status (driver): " +
+      Option(pod).map(formatPodState).getOrElse("unknown."))
+  }
+
+  override def executeOnGlob(pods: List[Pod], ns: String, gracePeriod: Option[Long])
+      (implicit client: KubernetesClient): Unit = {
+    if (pods.nonEmpty) {
+      for (pod <- pods) {
+        printMessage("Application status (driver): " +
+          Option(pod).map(formatPodState).getOrElse("unknown."))
+      }
+    } else {
+      printMessage("No applications found.")
+    }
+  }
+}
+
+private[spark] class K8sSparkSubmitOperationsClient extends SparkSubmitOperationsClient
+  with CommandLineLoggingUtils {
+  private def isGlob(name: String): Boolean = {
+    name.last == '*'
+  }
+
+  def execute(submissionId: String, sparkConf: SparkConf, op: K8sSubmitOp): Unit = {
+    val master = KubernetesUtils.parseMasterUrl(sparkConf.get("spark.master"))
+    val sIdParts = submissionId.split(":") match {
+      case Array(ns, pName) => Some((ns, pName))
+      case Array(pName) => Some(("default", pName))
```

Comment:
I swear I remember asking about this...

But if I submit an app to k8s, without specifying a namespace, is the namespace "default", or can it change depending on my k8s client config?

If the latter, this needs to behave accordingly.

Reply:
It is the default, I documented it.

Reply:
This is broken on minikube btw: https://github.com/kubernetes/minikube/issues/2117.
I will have to first load the config then check if namespace exists etc. Right now I just re-user the code at the backend, which will pick the .kub/conf config by default and then override it with whatever we pass.